### PR TITLE
Make translation of optimize statements dependent on literal count

### PR DESCRIPTION
**Pull request by MaxOstrowski**
Refer to issue #55.

@rkaminsk
The summing up the size of the minimize constraint
```
for (auto [co, var] : *minimize) {
            auto & vs = master_().var_state(var);
            min_size += static_cast<uint64_t>(vs.max_bound() - vs.min_bound() - 1);
        }
```
is done twice, which is ugly.
Should I move this to a seperate `solver.literals_in_objective()` function ?
Any suggestions ?

### CODE REVIEW

File: libclingcon/src/propagator.cc [hunk 1]
```diff
@@ -491,9 +497,11 @@ bool Propagator::translate_(InitClauseCreator &cc, UniqueMinimizeConstraint mini
     cc.set_state(InitState::Init);
 
     // mark minimize constraint as translated if necessary
-    if (config_.translate_minimize && minimize_ != nullptr) {
-        translated_minimize_ = true;
-        minimize_ = nullptr;
+    if (minimize_ != nullptr) {
+           if (config_.translate_minimize < 0 || min_size < static_cast<int64_t>(config_.translate_minimize)) {
```
Comment (rkaminsk) [inline libclingcon/src/propagator.cc:501]: wrong indentation and unnecessary if

Comment (MaxOstrowski) [reply libclingcon/src/propagator.cc]: hopefully fixed

**Comment by MaxOstrowski** [comment]
Moved function to MinimizeConstraint

**Comment by rkaminsk** [comment]
Comparing int32 and int64 does not need a cast. The int32_t should automatically be promoted to int64_t. Casts are only necessary if signed and unsigned integers are involved in a cast. Please check the logic when to translate. It looks like you want to have something like

- `translate_minimize < 0` always translate,
- `translate_minimize > 0` translate if smaller, and
- `translate_minimize == 0` do not translate.

Right now the condition to set `translated_minimize_` is not the exact opposite of what is used in the minimize constraint. And here is also where it gets tricky. Before it was easy enough to check if the constraint was removed or was kept depending on the option value (I believe the domain of a variable can change during translation). This means, now, you have to find a better way to check if the constraint is still in the vector*. Also before the algorithm would only throw an error if the user changed the value of the translate minimize option during incremental solving now it can also fail if in a first step a small enough minimize constraint is translated and then a too large one is added. How do you want to handle this?

(*) You could walk through the constraint vector to see if the minimize constraint is still in it (ugly because overhead). Or set a flag somewhere (ugly because it makes the code harder to maintain). Or change the way the minimize constraint is handled completely (might be involved and require some thinking).

**Comment by MaxOstrowski** [comment]
The idea for multi-shot would be:
Once a minimize statement is translated, all future minimize constraints are always translated.

**Comment by MaxOstrowski** [comment]
- [x] removed unecessary casts
- [x] replaced second condition check with flag
- [x] used flag to force translation if it was done once
- [ ] should we remove the error "translation of minimize constraints is disabled but was enabled before" ?

**Comment by rkaminsk** [comment]
It looks way better than before. I am a bit busy and have a closer look later.

**Comment by rkaminsk** [comment]
I refactored this a bit.
- I am using `uint32_t` to configure the translation option. Setting the value to the maximum guarantees that the minimize constraint will always be translated because it can only have `UINT_MAX` elements. (Once #62 is merged, it can also be conveniently be set using `max`).
- I moved the translated flag to the solver because I felt it fits better into the design.
- I removed the error message as you suggested.
@MaxOstrowski  please have a look if this is okay with you.

**Review by rkaminsk — APPROVED** [review]

**Review by MaxOstrowski — COMMENTED** [review]
Looks good. Thanks a lot.

**Pull request merged**